### PR TITLE
feat(rpc):transaction variant for all available transaction formats

### PR DESCRIPTION
**Pull request by supernovahs**
Ref #5093

### CODE REVIEW

File: crates/rpc/rpc/src/transactionvariant.rs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+/// Different transaction formats under one hood
+#[derive(Debug, Clone)]
+pub enum TransactionVariant {
+    /// Transaction variant
+    RegularTx(Transaction),
+    /// TransactionSigned variant
+    SignedTx(TransactionSigned),
+    /// TransactionSignedEcRecovered variant
+    SignedEcRecoveredTx(TransactionSignedEcRecovered),
```
Comment (mattsse) [inline crates/rpc/rpc/src/transactionvariant.rs:13]:
```suggestion
    /// Transaction variant
    Plain(Transaction),
    /// TransactionSigned variant
    Signed(TransactionSigned),
    /// TransactionSignedEcRecovered variant
    SignedEcRecovered(TransactionSignedEcRecovered),
```

File: crates/rpc/rpc/src/transactionvariant.rs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    /// Consumes the `TransactionVariant`
+    /// Returns the consumed `TransactionSignedEcRecovered ` if its a `SignedEcRecoveredTx` variant
+    /// Otherwise returns None
+    pub fn into_signed_ec_recovered_tx(self) -> Option<TransactionSignedEcRecovered> {
+        match self {
+            TransactionVariant::SignedEcRecoveredTx(tx) => Some(tx),
+            _ => None,
+        }
```
Comment (mattsse) [inline crates/rpc/rpc/src/transactionvariant.rs:86]: the into function need to convert them, by hashing and recovering if necessary
or extracting the wrapped component

File: crates/rpc/rpc/src/transactionvariant.rs
Comment (mattsse) [inline crates/rpc/rpc/src/transactionvariant.rs:1]: this should be moved to primitives transaction.rs

**Review by mattsse — CHANGES_REQUESTED** [review]
awesome, only a few suggestions

File: crates/primitives/src/transaction/transactionvariant.rs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            }
+            TransactionVariant::Plain(tx) => {
+                let mut instance =
+                    TransactionSigned { transaction: tx, signature, hash: Default::default() };
+                instance.hash = instance.recalculate_hash();
+                let signer = instance.recover_signer();
+                Some(TransactionSignedEcRecovered {
+                    signer: signer.unwrap(),
+                    signed_transaction: instance,
+                })
```
Comment (mattsse) [inline crates/primitives/src/transaction/transactionvariant.rs:107]: use existing conversion functions, see `fn into_` etc

File: crates/primitives/src/transaction/transactionvariant.rs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+                instance.hash = instance.recalculate_hash();
+                Some(instance)
+            }
+            TransactionVariant::SignedEcRecovered(tx) => Some(tx.signed_transaction),
+        }
+    }
+
+    /// Consumes the `TransactionVariant`
+    /// Returns the consumed `TransactionSignedEcRecovered `
+    pub fn into_signed_ec_recovered(
```
Comment (mattsse) [inline crates/primitives/src/transaction/transactionvariant.rs:86]: needs docs when this returns None

File: crates/primitives/src/transaction/transactionvariant.rs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    /// Consumes the `TransactionVariant`
+    /// Returns the consumed `TransactionSigned `
+    pub fn into_signed(self, signature: Signature) -> Option<TransactionSigned> {
+        match self {
+            TransactionVariant::Signed(tx) => Some(tx),
+            TransactionVariant::Plain(tx) => {
+                let mut instance =
+                    TransactionSigned { transaction: tx, signature, hash: Default::default() };
+                instance.hash = instance.recalculate_hash();
+                Some(instance)
```
Comment (mattsse) [inline crates/primitives/src/transaction/transactionvariant.rs:78]: same

File: crates/primitives/src/transaction/transactionvariant.rs [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        match self {
+            TransactionVariant::Plain(tx) => Some(tx),
+            TransactionVariant::Signed(tx) => Some(tx.transaction),
+            TransactionVariant::SignedEcRecovered(tx) => Some(tx.signed_transaction.transaction),
+        }
+    }
+
+    /// Consumes the `TransactionVariant`
+    /// Returns the consumed `TransactionSigned `
+    pub fn into_signed(self, signature: Signature) -> Option<TransactionSigned> {
```
Comment (mattsse) [inline crates/primitives/src/transaction/transactionvariant.rs:71]: also needs docs when this returns none

File: crates/primitives/src/transaction/transactionvariant.rs [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+/// Different transaction formats under one hood
+#[derive(Debug, Clone)]
+pub enum TransactionVariant {
+    /// Transaction variant
+    Plain(Transaction),
+    /// TransactionSigned variant
+    Signed(TransactionSigned),
+    /// TransactionSignedEcRecovered variant
+    SignedEcRecovered(TransactionSignedEcRecovered),
+}
```
Comment (mattsse) [inline crates/primitives/src/transaction/transactionvariant.rs:14]: we also want `From<>` impls for this type for all variants

**Review by mattsse — CHANGES_REQUESTED** [review]
almost there

File: crates/primitives/src/transaction/transactionvariant.rs (hunk 6 above)
Comment (supernovahs) [reply crates/primitives/src/transaction/transactionvariant.rs:71]: I don't see any case when None will be returned?

**Review by mattsse — APPROVED** [review]
nice work.

while reviewing this, my initial assumption that `Plain(Transaction)` is useful was wrong, because we always have a Plain+Signature, that's how we store it in the db.

so it makes more sense to always keep the signature, which we need, hence I added a SignedNoHash variant and replaced the Plain variant.

all of them can be downcasted to a `Transaction anyway`

**Pull request merged**